### PR TITLE
Save serial log after testcase completion 

### PR DESCRIPTION
**Pull request by adityagesh**
Save serial log of VMs after testcase completes.

Example log file path: 
1. runtime\log\20230410\20230410-095213-491\environments\20230410-095223-846-generated_0\lisa-remote1234-20230410-095213-491-e0-n0
2. runtime\log\20230410\20230410-095213-491\environments\20230410-095223-846-generated_0\lisa-remote1234-20230410-095213-491-e0-n1

### CODE REVIEW

**Comment by squirrelsc** [comment]
When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.

**Comment by LiliDeng** [comment]
> When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.

in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)

is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?

**Comment by squirrelsc** [comment]
> > When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.
> 
> in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)
> 
> is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?

That can be collected in test cases by calling the feature. Make it configurable is not easy to use than add one line code.

**Comment by adityagesh** [comment]
> > > When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.
> > 
> > 
> > in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)
> > is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?
> 
> That can be collected in test cases by calling the feature. Make it configurable is not easy to use than add one line code.

I can think of two cases where we would like to have the serial logs
1. Perf tests where test passes but perf numbers are low
2. Failed cases - sometimes failures pass on rerun. It is good to have serial logs in these cases.

**Comment by squirrelsc** [comment]
> > > > When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.
> > > 
> > > 
> > > in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)
> > > is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?
> > 
> > 
> > That can be collected in test cases by calling the feature. Make it configurable is not easy to use than add one line code.
> 
> I can think of two cases where we would like to have the serial logs
> 
> 1. Perf tests where test passes but perf numbers are low
> 2. Failed cases - sometimes failures pass on rerun. It is good to have serial logs in these cases.

Can we collect log before the environments deleted? I think collect one time for each env should be enough. The serial log contains timestamp too. We can know what happened during each test cases from this log.

**Comment by adityagesh** [comment]
> > > > > When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.
> > > > 
> > > > 
> > > > in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)
> > > > is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?
> > > 
> > > 
> > > That can be collected in test cases by calling the feature. Make it configurable is not easy to use than add one line code.
> > 
> > 
> > I can think of two cases where we would like to have the serial logs
> > 
> > 1. Perf tests where test passes but perf numbers are low
> > 2. Failed cases - sometimes failures pass on rerun. It is good to have serial logs in these cases.
> 
> Can we collect log before the environments deleted? I think collect one time for each env should be enough. The serial log contains timestamp too. We can know what happened during each test cases from this log.

@johnsongeorge-w , Please share your thoughts on this suggestion

**Comment by johnsongeorge-w** [comment]
> > > > > > When is it needed? It costs time to capture on each test case. If some test cases need it, it can be called in the test cases.
> > > > > 
> > > > > 
> > > > > in v2. We collect dmesg when testing starts and ends. now in v3, at the end of test, we don't collect such info since it uses the cache. sometimes we want to check there is any call trace or errors shown up in log (the case may pass or fail)
> > > > > is it feasible that we add a switch to enable collect log at end of testing, by default, it is set as false?
> > > > 
> > > > 
> > > > That can be collected in test cases by calling the feature. Make it configurable is not easy to use than add one line code.
> > > 
> > > 
> > > I can think of two cases where we would like to have the serial logs
> > > 
> > > 1. Perf tests where test passes but perf numbers are low
> > > 2. Failed cases - sometimes failures pass on rerun. It is good to have serial logs in these cases.
> > 
> > 
> > Can we collect log before the environments deleted? I think collect one time for each env should be enough. The serial log contains timestamp too. We can know what happened during each test cases from this log.
> 
> @johnsongeorge-w , Please share your thoughts on this suggestion

There are 2 major use case and I think few aspects are already discussed above.
1.	Collecting dmesg before and after each test will help to narrow down the potential problem (from kernel error message perspective) when analyzing any test failure or performance issues.
2.	Serial log is equally important to be collected after each failed test cases. In many situations, we are not able to conclude because lack of serial logs. Checking **last bootup (only for failed tests) ** will be more convenient rather searching the entire serial logs.  

Collecting these logs will considerably optimize (avoid re-runs) the time required to triage/analyze the test failures. This is critical when issue is intermittent and cannot be reproduced always.

File: lisa/testsuite.py [hunk 1]
```diff
@@ -640,6 +640,13 @@ def start(
                 log=case_log,
             )
 
+            if environment.platform:
```
Comment (squirrelsc) [inline lisa/testsuite.py:643]: 1. Never call private method out of the class. It breaks the design and fail other platform.
2. In test cases, use node.features.is_supported to check if SerialConsole is available, and then call it. Don't call platform directly.

File: lisa/platform_.py [hunk 2]
```diff
@@ -107,6 +107,14 @@ def _cleanup(self) -> None:
         """
         pass
 
+    def _save_console_log(
```
Comment (squirrelsc) [inline lisa/platform_.py:110]: It's a feature of the node, not add it to platform.

**Pull request closed without merging**